packages/@lwc/engine-core/src/framework/utils.ts

```diff
@@ -98,3 +100,20 @@ export function flattenStylesheets(stylesheets: TemplateStylesheetFactories): St
     }
     return list;
 }
+
+// Set a ref (lwc:ref) on a VM, from a template API
+export function setRefVNode(vm: VM, ref: string | undefined, vnode: VNode) {
+    if (!isUndefined(ref)) {
+        // Create the refVNodes object on-demand. This ensures that templates with no refs
+        // don't pay the perf tax of creating objects unnecessarily.
+        const refVNodes: RefVNodes = isNull(vm.refVNodes)
+            ? (vm.refVNodes = create(null))
+            : vm.refVNodes;
+
+        // In cases of conflict (two elements with the same ref), prefer, the last one,
+        // in depth-first traversal order.
+        if (!(ref in refVNodes) || refVNodes[ref].key! < vnode.key!) {
```

Comment:
`VBaseElement`s always have keys. See proposed change above.

```suggestion
        if (!(ref in refVNodes) || refVNodes[ref].key < vnode.key) {
```